File: experiments/control-run/go/stats.go
package main

import (
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"
)

const (
	defaultTop = 5
	minTop     = 1
	maxTop     = 100
)

// ErrContract is wrapped by every broken requirement or invariant.
var ErrContract = errors.New("contract broken")

func broken(format string, args ...any) error {
	return fmt.Errorf("%w: %s", ErrContract, fmt.Sprintf(format, args...))
}

// Summary accumulates records one at a time so no file is held in memory.
// It keeps only the top-N slowest and one counter per method and path.
type Summary struct {
	top       int
	requests  int
	errors    int
	successes int
	malformed int
	first     time.Time
	last      time.Time
	seq       int
	slowest   []ranked
	counts    map[endpoint]int
}

type ranked struct {
	Record
	seq int
}

type endpoint struct {
	Method string
	Path   string
}

// PathCount is one row of the busiest list.
type PathCount struct {
	endpoint
	Count int
}

// Result is the finished summary, ready to render.
type Result struct {
	Requests  int
	Errors    int
	Malformed int
	ErrorRate float64
	PerMinute float64
	Slowest   []Record
	Busiest   []PathCount
}

// NewSummary requires top in 1..100: out of range is an error, not a clamp.
func NewSummary(top int) (*Summary, error) {
	if err := CheckTop(top); err != nil {
		return nil, err
	}
	return &Summary{top: top, counts: map[endpoint]int{}}, nil
}

// CheckTop requires n in 1..100.
func CheckTop(n int) error {
	if n < minTop || n > maxTop {
		return broken("top %d outside %d..%d", n, minTop, maxTop)
	}
	return nil
}

// AddMalformed counts one line that did not parse.
func (s *Summary) AddMalformed() { s.malformed++ }

// Add requires a record ParseLine could have produced; it refuses any
// other without changing the summary.
func (s *Summary) Add(r Record) error {
	if err := checkRecord(r); err != nil {
		return err
	}
	s.count(r)
	s.span(r.At)
	s.rankSlow(ranked{Record: r, seq: s.seq})
	s.seq++
	return nil
}

func checkRecord(r Record) error {
	switch {
	case r.Status < 100 || r.Status > 599:
		return broken("status %d outside 100..599", r.Status)
	case ContainsCard(r.Path) || ContainsCard(r.AtText):
		return broken("record holds an unmasked card number")
	case checkMethod(r.Method) != nil || checkPath(r.Path) != nil:
		return broken("record method or path not well formed")
	}
	return nil
}

func (s *Summary) count(r Record) {
	s.requests++
	if r.Status >= 500 {
		s.errors++
	} else {
		s.successes++
	}
	s.counts[endpoint{r.Method, r.Path}]++
}

func (s *Summary) span(at time.Time) {
	if s.requests == 1 || at.Before(s.first) {
		s.first = at
	}
	if s.requests == 1 || at.After(s.last) {
		s.last = at
	}
}

// slower orders the slowest list: duration descending, then timestamp
// ascending, then arrival order, so ties are stable and deterministic.
func slower(a, b ranked) bool {
	if a.Ms != b.Ms {
		return a.Ms > b.Ms
	}
	if !a.At.Equal(b.At) {
		return a.At.Before(b.At)
	}
	return a.seq < b.seq
}

func (s *Summary) rankSlow(r ranked) {
	i := sort.Search(len(s.slowest), func(i int) bool { return slower(r, s.slowest[i]) })
	if i >= s.top {
		return
	}
	s.slowest = append(s.slowest, ranked{})
	copy(s.slowest[i+1:], s.slowest[i:])
	s.slowest[i] = r
	if len(s.slowest) > s.top {
		s.slowest = s.slowest[:s.top]
	}
}

// Check holds the summary's nevers: requests = errors + successes and
// errors <= requests.
func (s *Summary) Check() error {
	if s.requests != s.errors+s.successes {
		return broken("requests %d != errors %d + successes %d", s.requests, s.errors, s.successes)
	}
	if s.errors > s.requests || s.errors < 0 {
		return broken("errors %d not within requests %d", s.errors, s.requests)
	}
	return nil
}

// Result finishes the summary, checking its invariants first.
func (s *Summary) Result() (Result, error) {
	if err := s.Check(); err != nil {
		return Result{}, err
	}
	return Result{
		Requests:  s.requests,
		Errors:    s.errors,
		Malformed: s.malformed,
		ErrorRate: ratio(s.errors, s.requests),
		PerMinute: s.perMinute(),
		Slowest:   s.slowRecords(),
		Busiest:   s.busiest(),
	}, nil
}

func ratio(part, whole int) float64 {
	if whole == 0 {
		return 0
	}
	return float64(part) / float64(whole)
}

// perMinute is requests over the span between the earliest and latest
// timestamp; a zero span (one request, or all at one instant) is 0.
func (s *Summary) perMinute() float64 {
	minutes := s.last.Sub(s.first).Minutes()
	if s.requests == 0 || minutes <= 0 {
		return 0
	}
	return float64(s.requests) / minutes
}

func (s *Summary) slowRecords() []Record {
	out := make([]Record, 0, len(s.slowest))
	for _, r := range s.slowest {
		out = append(out, r.Record)
	}
	return out
}

// busiest sorts by count descending, then path ascending, then method
// ascending, and keeps the top N.
func (s *Summary) busiest() []PathCount {
	rows := make([]PathCount, 0, len(s.counts))
	for e, n := range s.counts {
		rows = append(rows, PathCount{endpoint: e, Count: n})
	}
	sort.Slice(rows, func(i, j int) bool { return busier(rows[i], rows[j]) })
	if len(rows) > s.top {
		rows = rows[:s.top]
	}
	return rows
}

func busier(a, b PathCount) bool {
	if a.Count != b.Count {
		return a.Count > b.Count
	}
	if c := strings.Compare(a.Path, b.Path); c != 0 {
		return c < 0
	}
	return a.Method < b.Method
}

File: experiments/control-run/python/stats.py
"""Tallying records into a Summary: counts, rate, the slowest requests and the busiest paths."""

from __future__ import annotations

import heapq
from dataclasses import dataclass, field

from contracts import require
from parse import Record, contains_card

TOP_MIN = 1
TOP_MAX = 100
ERROR_MIN = 500
ERROR_MAX = 599
_MICROSECONDS_PER_MINUTE = 60_000_000


@dataclass(frozen=True, slots=True)
class Busy:
    """How many requests one method and path received."""

    count: int
    method: str
    path: str

    def __post_init__(self) -> None:
        require(self.count >= 1, "count is at least 1")
        require(not contains_card(self.path), "path holds no card number")


@dataclass(frozen=True, slots=True)
class Summary:
    requests: int
    errors: int
    successes: int
    malformed: int
    per_minute: float
    slowest: tuple[Record, ...]
    busiest: tuple[Busy, ...]

    def __post_init__(self) -> None:
        require(self.requests == self.errors + self.successes, "requests equals errors + successes")
        require(0 <= self.errors <= self.requests, "errors <= requests")
        require(self.malformed >= 0 and self.per_minute >= 0.0, "malformed and per_minute are not negative")
        require(_ascending([slow_key(r) for r in self.slowest]), "slowest by duration descending, then timestamp ascending")
        require(_ascending([busy_key(b) for b in self.busiest]), "busiest by count descending, then path ascending")

    @property
    def error_rate(self) -> float:
        """Errors over requests; 0.0 when there are no requests."""
        return self.errors / self.requests if self.requests else 0.0


def is_error(status: int) -> bool:
    return ERROR_MIN <= status <= ERROR_MAX


def slow_key(record: Record) -> tuple[int, int]:
    return (-record.duration_ms, record.instant)


def busy_key(busy: Busy) -> tuple[int, str, str]:
    return (-busy.count, busy.path, busy.method)


def _ascending(keys: list[tuple[int, int]] | list[tuple[int, str, str]]) -> bool:
    return all(earlier <= later for earlier, later in zip(keys, keys[1:], strict=False))


@dataclass(slots=True)
class Tally:
    """Accumulates records one at a time; memory is bounded by `top` plus the distinct method-path pairs."""

    top: int
    requests: int = 0
    errors: int = 0
    successes: int = 0
    malformed: int = 0
    first: int | None = None
    last: int | None = None
    # min-heap whose root is the least slow of the kept records: (ms, -instant, -arrival, record)
    _slowest: list[tuple[int, int, int, Record]] = field(default_factory=list)
    _counts: dict[tuple[str, str], int] = field(default_factory=dict)

    def __post_init__(self) -> None:
        require(TOP_MIN <= self.top <= TOP_MAX, "top is 1 to 100")

    def add(self, record: Record) -> None:
        self.requests += 1
        if is_error(record.status):
            self.errors += 1
        else:
            self.successes += 1
        self.first = record.instant if self.first is None else min(self.first, record.instant)
        self.last = record.instant if self.last is None else max(self.last, record.instant)
        self._offer_slow(record)
        key = (record.method, record.path)
        self._counts[key] = self._counts.get(key, 0) + 1

    def add_malformed(self) -> None:
        self.malformed += 1

    def summary(self) -> Summary:
        return Summary(
            requests=self.requests,
            errors=self.errors,
            successes=self.successes,
            malformed=self.malformed,
            per_minute=self._per_minute(),
            slowest=self._ranked_slowest(),
            busiest=self._ranked_busiest(),
        )

    def _offer_slow(self, record: Record) -> None:
        """Keep the `top` slowest; ties go to the earlier timestamp, then to the earlier arrival."""
        entry = (record.duration_ms, -record.instant, -self.requests, record)
        if len(self._slowest) < self.top:
            heapq.heappush(self._slowest, entry)
        elif entry[:3] > self._slowest[0][:3]:
            heapq.heapreplace(self._slowest, entry)

    def _per_minute(self) -> float:
        if self.first is None or self.last is None or self.first == self.last:
            return 0.0
        return self.requests * _MICROSECONDS_PER_MINUTE / (self.last - self.first)

    def _ranked_slowest(self) -> tuple[Record, ...]:
        ranked = sorted(self._slowest, key=lambda entry: entry[:3], reverse=True)
        return tuple(entry[3] for entry in ranked)

    def _ranked_busiest(self) -> tuple[Busy, ...]:
        ranked = heapq.nsmallest(
            self.top, self._counts.items(), key=lambda item: (-item[1], item[0][1], item[0][0])
        )
        return tuple(Busy(count=count, method=method, path=path) for (method, path), count in ranked)
